File: review/eval/r2-p6/runs/2026-09-24-round3/work/timebox-two-modules.A.1/ledger.py
class Ledger:
    def apply(self, rows, account):
        result = []
        for row in rows:
            kind, acct, cents = row
            if cents < 0:
                raise ValueError("negative amount: %r" % (cents,))
            if acct == account:
                result.append((kind, acct, cents))
        return result


def balance(rows, account):
    total = 0
    for kind, acct, cents in rows:
        if acct != account:
            continue
        if kind == "deposit":
            total += cents
        elif kind == "withdraw":
            total -= cents
    return total

File: review/eval/archive/tasks/rust-ledger/fixture/src/ledger.rs
use std::collections::BTreeMap;
use crate::csv::Event;

#[derive(Debug, Default)]
pub struct Ledger {
    balances: BTreeMap<String, i64>,
    requests: BTreeMap<String, Event>,
}

impl Ledger {
    pub fn new() -> Self { Self::default() }

    pub fn balance(&self, account: &str) -> Option<i64> { self.balances.get(account).copied() }

    pub fn balances(&self) -> &BTreeMap<String, i64> { &self.balances }

    pub fn apply(&mut self, event: &Event) -> Result<bool, String> {
        if self.requests.contains_key(event.id()) { return Ok(false); }
        self.requests.insert(event.id().to_owned(), event.clone());
        match event {
            Event::Deposit { account, cents, .. } => {
                *self.balances.entry(account.clone()).or_default() += cents;
            }
            Event::Transfer { from, to, cents, .. } => {
                let source = self.balances.get_mut(from).ok_or("源账户不存在")?;
                if *source < *cents { return Err("余额不足".to_owned()); }
                *source -= cents;
                let target = self.balances.get_mut(to).ok_or("目标账户不存在")?;
                *target += cents;
            }
        }
        Ok(true)
    }
}
